pkg/providers/version/version.go

```diff
@@ -30,35 +36,94 @@ const (
 	kubernetesVersionCacheKey = "kubernetesVersion"
 )
 
+type HTTPClient interface {
+	Get(string) (*http.Response, error)
+}
+
 // Provider get the APIServer version. This will be initialized at start up and allows karpenter to have an understanding of the cluster version
 // for decision making. The version is cached to help reduce the amount of calls made to the API Server
-
 type Provider struct {
-	cache               *cache.Cache
-	cm                  *pretty.ChangeMonitor
-	kubernetesInterface kubernetes.Interface
+	cache      *cache.Cache
+	cm         *pretty.ChangeMonitor
+	kubeClient client.Client
+	httpClient HTTPClient
 }
 
-func NewProvider(kubernetesInterface kubernetes.Interface, cache *cache.Cache) *Provider {
+func NewProvider(cache *cache.Cache, client client.Client, httpClient HTTPClient) *Provider {
 	return &Provider{
-		cm:                  pretty.NewChangeMonitor(),
-		cache:               cache,
-		kubernetesInterface: kubernetesInterface,
+		cm:         pretty.NewChangeMonitor(),
+		cache:      cache,
+		kubeClient: client,
+		httpClient: httpClient,
 	}
 }
 
 func (p *Provider) Get(ctx context.Context) (string, error) {
 	if version, ok := p.cache.Get(kubernetesVersionCacheKey); ok {
 		return version.(string), nil
 	}
-	serverVersion, err := p.kubernetesInterface.Discovery().ServerVersion()
+	version, err := p.GetMinKubernetesVersion(ctx)
 	if err != nil {
 		return "", err
 	}
-	version := fmt.Sprintf("%s.%s", serverVersion.Major, strings.TrimSuffix(serverVersion.Minor, "+"))
 	p.cache.SetDefault(kubernetesVersionCacheKey, version)
 	if p.cm.HasChanged("kubernetes-version", version) {
 		logging.FromContext(ctx).With("version", version).Debugf("discovered kubernetes version")
 	}
 	return version, nil
 }
+
+// GetMinKubernetesVersion ensures that we query all known APIServers for the K8s version.
+// This to handle any scenarios where there may be multiple APIServer reporting different
+// K8s versions.
+func (p *Provider) GetMinKubernetesVersion(ctx context.Context) (string, error) {
+	var endpointSlice v1.EndpointSlice
+	if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: "kubernetes", Namespace: "default"}, &endpointSlice); err != nil {
+		return "", fmt.Errorf("getting endpoints, %w", err)
+	}
+	// This error should never occur
+	minVersion, err := version.ParseGeneric("v99.99.99")
```

Comment:
Is it ever possible that the endpoint slice returns an empty list? In that case, we are returning this version. It seems like in that case, we should error for safety, saying something like "failed to discover the version since we didn't get any apiserver endpoints"

Reply:
I think there'd have to be no APIServer endpoints, but you bring up a good point. Let me fix this